### PR TITLE
Add approval check for infrabuild and deploy to environments

### PR DESCRIPTION
**Pull request by njrich28**

### CODE REVIEW

File: src/uk/gov/hmcts/pipeline/EnvironmentApprovals.groovy [hunk 1]
```diff
@@ -0,0 +1,37 @@
+package uk.gov.hmcts.pipeline
+
+class EnvironmentApprovals {
+  def steps
+  static final String GITHUB_CREDENTIAL = 'jenkins-github-hmcts-api-token'
+  static def approvalsMap
```
Comment (timja) [inline src/uk/gov/hmcts/pipeline/EnvironmentApprovals.groovy]: shouldn't need to include the type in a variable name?
if you're tempted to do that and IDE can't do it for you maybe put the actual type in groovy?
I assume it's

```suggestion
  private static Map<String, Object> approvalsMap
```

File: vars/sectionDeployToEnvironment.groovy [hunk 2]
```diff
@@ -21,66 +21,68 @@ def call(params) {
   Deployer deployer = pipelineType.deployer
   def tfOutput
 
-  lock(resource: "${product}-${component}-${environment}-deploy", inversePrecedence: true) {
-    folderExists('infrastructure') {
-      stage("Build Infrastructure - ${environment}") {
-        onPreview {
-          deploymentNumber = githubCreateDeployment()
-        }
+  approvedEnvironmentReposiitory(environment) {
```
Comment (timja) [inline vars/sectionDeployToEnvironment.groovy]:
```suggestion
  approvedEnvironmentRepository(environment) {
```

File: src/uk/gov/hmcts/pipeline/EnvironmentApprovals.groovy [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+
+  boolean isApproved(String environment, String repository) {
+    def environmentApprovalsMap = getEnvironmentApprovalsMap()
+    if (!environmentApprovalsMap.containsKey(environment)) {
+      // if environment is not listed in config then assume no whitelist required
+      return true
+    }
+
+    def approvals = environmentApprovalsMap.get(environment)
+    return approvals.findResult(false) { it.repo.equalsIgnoreCase(repository) ? true : null }
```
Comment (timja) [inline src/uk/gov/hmcts/pipeline/EnvironmentApprovals.groovy]: If we do this it's going to be a complete breaking change.
few options I guess,
1. just break it
2. Implement SlackNagger (worthwhile imo)
3. do a best efforts pre-fill it with what's currently in prod and break it

File: src/uk/gov/hmcts/pipeline/EnvironmentApprovals.groovy [hunk 4]
```diff
@@ -0,0 +1,37 @@
+package uk.gov.hmcts.pipeline
+
+class EnvironmentApprovals {
```
Comment (timja) [inline src/uk/gov/hmcts/pipeline/EnvironmentApprovals.groovy]: Probably worth a test? Should be able to lift most of it from the TeamNames test?

File: src/uk/gov/hmcts/pipeline/EnvironmentApprovals.groovy (hunk 1 above)
Comment (njrich28) [reply src/uk/gov/hmcts/pipeline/EnvironmentApprovals.groovy]: cut'n'paste from TeamNames 😄 - probably better called environmentApprovals anyway

File: src/uk/gov/hmcts/pipeline/EnvironmentApprovals.groovy (hunk 3 above)
Comment (njrich28) [reply src/uk/gov/hmcts/pipeline/EnvironmentApprovals.groovy]: So, start with 2 (I'll actually change to make it non-breaking) and warn/notify people
Then 3
Then 1

File: vars/approvedEnvironmentRepository.groovy [hunk 5]
```diff
@@ -10,9 +10,23 @@ import uk.gov.hmcts.pipeline.EnvironmentApprovals
  * }
  */
 def call(String environment, Closure block) {
-  if (new EnvironmentApprovals(this).isApproved(environment, env.GIT_URL)) {
-    return block.call()
-  } else {
-    echo "Repo ${env.GIT_URL} is not approved for environment ${environment}"
+  if (!new EnvironmentApprovals(this).isApproved(environment, env.GIT_URL)) {
+    echo '''
+================================================================================
```
Comment (timja) [inline vars/approvedEnvironmentRepository.groovy]: gonna do the slack nagger as well?

**Pull request merged**